jax/experimental/loops.py

```diff
@@ -0,0 +1,513 @@
+                    arr1,
+                    lambda arr1: arr1)
+  arr = lax.fori_loop(0, arr.shape[0], loop_body, arr)
+
+The default notation quickly gets unreadable with deeper nested loops.
+With the utilities in this module you can write loops and conditionals that look closer to plain Python,
+as long as you keep the loop-carried state in a special `loops.scope` object and use `for` loops
+over special `scope.range` iterators::
+
+  from jax.experimental import loops
+  with loops.Scope() as s:
+    s.arr = np.zeros(5)  # Must create the mutable state of the loop as `scope` fields.
+    for i in s.range(s.arr.shape[0]):
+      s.arr = ops.index_update(s.arr, i, s.arr[i] + 2.)
+      for _ in s.cond_range(i % 2 == 0):  # Conditionals are also sugared as loops with 0 or 1 iterations
+        s.arr = ops.index_update(s.arr, i, s.arr[i] + 1.)
+
+Notes:
+  * Loops and conditionals to be functionalized can appear only inside scopes constructed with `loops.Scope`
+    and they must use one of the `Scope.range` iterators. All other loops are unrolled during tracing, as usual in JAX.
+  * Only scope data (stored in fields of the scope object) is functionalized. All other state, e.g., in other
+    Python variables, will not be considered as being part of the loop output.
+    All references to the mutable state should be through the scope: `s.arr`.
+  * Conceptually, this model is still "functional" in the sense that a loop over a `Scope.range` behaves
+    as a function whose input and output is the scope data.
+  * Scopes should be passed down to callees that need to use loop functionalization, or they may be nested.
+  * The programming model is that the loop body over a `scope.range` is traced only once, using abstract shape values,
+    similar to how JAX traces function bodies.
+
+Restrictions:
+  * The tracing of the loop body should not exit prematurely with `return`, `exception`, `break`.
+    This would be detected and reported as errors when we encounter unnested scopes.
+  * The loop index variable should not be used after the loop. Similarly, one should not use outside the loop
+    data computed in the loop body, except data stored in fields of the scope object.
+  * No new mutable state can be created inside a loop to be functionalized. All mutable state must be created outside
+    all loops and conditionals.
+  * For a `while` loop, the conditional function is not allowed to modify the scope state. This is a checked error.
+    Also, for `while` loops the `grad` transformation does not work. An alternative that allows `grad` is a bounded
+    loop (`range`) with a conditional to ignore a suffix of iterations.
+
+Transformations:
+  * All transformations are supported, except `grad` is not supported for `Scope.while_range` loops.
+  * `vmap` is very useful for such loops because it pushes more work into the inner-loops, which should help
+    performance for accelerators.
+
+For usage example, see tests/loops_test.py.
```

Comment:
I would suggest copying in a few test cases into the docstring here. In particular it would be great to have at least one `while` loop example.

Reply:
Added an example with while_rage.